ansible_rulebook/job_template_runner.py

```diff
@@ -0,0 +1,180 @@
+
+import aiohttp
+import dpath
+
+from ansible_rulebook.exception import ControllerApiException
+
+logger = logging.getLogger(__name__)
+
+
+# https://docs.ansible.com/ansible-tower/latest/html/towerapi/access_resources.html#access-resources
+NAMED_URL_RES_INNER_DILIMITER = "-"
+URL_PATH_RESERVED_CHARSET = {}
+for c in ";/?:@=&[]":
+    URL_PATH_RESERVED_CHARSET[c] = quote(c, safe="")
+URL_PATH_RESERVED_CHARSET["+"] = "[+]"
+
+
+def _encode_uri(text):
+    for c in URL_PATH_RESERVED_CHARSET:
+        if c in text:
+            text = text.replace(c, URL_PATH_RESERVED_CHARSET[c])
+    text = text.replace(
+        NAMED_URL_RES_INNER_DILIMITER, "[%s]" % NAMED_URL_RES_INNER_DILIMITER
+    )
+    return text
+
+
+class JobTemplateRunner:
+    JOB_TEMPLATE_SLUG = "/api/v2/job_templates"
+    VALID_POST_CODES = [200, 201, 202]
+    JOB_COMPLETION_STATUSES = ["successful", "failed", "error", "canceled"]
+
+    def __init__(self, host: str, token: str):
+        self.token = token
+        self.host = host
+        self.refresh_delay = int(
+            os.environ.get("EDA_JOB_TEMPLATE_REFRESH_DELAY", 10)
+        )
+
+    async def _get_page(self, session, href_slug, params):
+        url = urljoin(self.host, href_slug)
+        async with session.get(url, params=params) as response:
+            response_text = dict(
+                status=response.status, body=await response.text()
+            )
+        if response_text["status"] != 200:
+            raise ControllerApiException(
+                "Failed to get from %s. Status: %s, Body: %s"
+                % (
+                    url,
+                    response_text["status"],
+                    response_text.get("body", "empty"),
+                )
+            )
+        return response_text
+
+    def _auth_headers(self):
+        return dict(Authorization=f"Bearer {self.token}")
+
+    async def run_job_template(
+        self,
+        name: str,
+        organization: str,
+        job_params: dict,
+        event_handler: Callable[[dict], Any],
+    ) -> str:
+        job = await self.launch(name, organization, job_params)
+
+        url_info = urlparse(job["url"])
+        url = f"{url_info.path}/job_events/"
+        counters = []
+        params = dict(parse_qsl(url_info.query))
+
+        async with aiohttp.ClientSession(
+            headers=self._auth_headers()
+        ) as session:
+            while True:
+                response = await self._get_page(session, url, params)
+                json_body = json.loads(response["body"])
+                job_status = None
+                for event in json_body["results"]:
+                    job_status = dpath.get(
+                        event, "summary_fields.job.status", "."
+                    )
+                    counter = event["counter"]
+                    if counter not in counters:
+                        counters.append(counter)
+                        print(event["stdout"])
```

Comment:
This should use the logger, don't?

Reply:
No. This is stdout, suppose to be printed to console.